backend/src/org/commcare/resources/model/installers/OfflineUserRestoreInstaller.java

```diff
@@ -0,0 +1,58 @@
+package org.commcare.resources.model.installers;
+
+import org.commcare.resources.model.Resource;
+import org.commcare.resources.model.ResourceInitializationException;
+import org.commcare.resources.model.ResourceLocation;
+import org.commcare.resources.model.ResourceTable;
+import org.commcare.resources.model.UnresolvedResourceException;
+import org.commcare.suite.model.OfflineUserRestore;
+import org.commcare.util.CommCareInstance;
+import org.javarosa.core.reference.Reference;
+import org.javarosa.xml.util.UnfullfilledRequirementsException;
+
+import java.io.IOException;
+
+/**
+ * Install user restore xml file present in app for use in offline logins.
+ * Used for providing a demo user restore.
+ *
+ * @author Phillip Mates (pmates@dimagi.com)
+ */
+public class OfflineUserRestoreInstaller extends CacheInstaller<OfflineUserRestore> {
+    @Override
+    protected String getCacheKey() {
+        return OfflineUserRestore.STORAGE_KEY;
+    }
+
+    @Override
+    public boolean initialize(CommCareInstance instance, boolean isUpgrade)
+            throws ResourceInitializationException {
+        instance.registerDemoUserRestore(storage().read(cacheLocation));
+        return true;
+    }
+
+    @Override
+    public boolean requiresRuntimeInitialization() {
+        return true;
+    }
+
+    @Override
+    public boolean install(Resource r, ResourceLocation location,
+                           Reference ref, ResourceTable table,
+                           CommCareInstance instance, boolean upgrade)
+            throws UnresolvedResourceException, UnfullfilledRequirementsException {
+        try {
+            OfflineUserRestore offlineUserRestore = OfflineUserRestore.buildInMemoryUserRestore(ref.getStream());
```

Comment:
Where are we catching exceptions for invalid file contents and things like that? Currently it feels like they'll fail over by throwing generic exceptions that look like IO exceptions or other network traffic problems